test/test_dtype.py

```diff
@@ -1,11 +1,29 @@
 import unittest
 import numpy as np
-from tinygrad.helpers import getenv, DType, DEBUG, ImageDType, PtrDType
+from tinygrad.helpers import CI, DTYPES_DICT, getenv, DType, DEBUG, ImageDType, PtrDType
 from tinygrad.ops import Device
 from tinygrad.tensor import Tensor, dtypes
-from typing import List, Optional
+from typing import Any, List
 from extra.utils import OSX, temp
-import copy
+
+def is_dtype_supported(dtype: DType):
+  # for GPU, cl_khr_fp16 isn't supported (except now we don't need it!)
+  # for LLVM, it segfaults because it can't link to the casting function
+  if dtype == dtypes.half: return not (CI and Device.DEFAULT in ["GPU", "LLVM"]) and Device.DEFAULT != "WEBGPU" and getenv("CUDACPU") != 1
+  if dtype == dtypes.bfloat16: return False # numpy doesn't support bf16, tested separately in TestBFloat16DType
+  if dtype == dtypes.float64: return Device.DEFAULT not in ["WEBGPU", "METAL"] and not OSX
+  if dtype in [dtypes.int8, dtypes.uint8]: return Device.DEFAULT not in ["WEBGPU"]
+  if dtype in [dtypes.int16, dtypes.uint16]: return Device.DEFAULT not in ["WEBGPU", "TORCH"]
+  if dtype == dtypes.uint32: return Device.DEFAULT not in ["TORCH"]
+  if dtype in [dtypes.int64, dtypes.uint64]: return Device.DEFAULT not in ["WEBGPU", "TORCH"]
+  if dtype == dtypes.bool:
+   # host-shareablity is a requirement for storage buffers, but 'bool' type is not host-shareable
+    if Device.DEFAULT == "WEBGPU": return False
+   # TODO remove triton from here once internal casting is fixed. CAST of fp32s between 0-1 is broken in triton
```

Comment:
[fixing triton](https://github.com/tinygrad/tinygrad/actions/runs/6683942229/job/18160671473#step:15:201) isn't a one-liner like clang 

It requires #2046 - The kernel that gets generated for this ast `Tensor([0.1,0.2,0.3,0.4], dtype=dtypes.float32).cast(dtypes.bool` needs to move from implicit store that truncates the input:
```
@triton.jit
def E_4(data0,data1):
  val0 = tl.load(data1 + 0, mask = True).to(tl.float32)
  val1 = tl.load(data1 + 1, mask = True).to(tl.float32)
  val2 = tl.load(data1 + 2, mask = True).to(tl.float32)
  val3 = tl.load(data1 + 3, mask = True).to(tl.float32)
  tl.store(data0 + 0, val0, mask = True) 
  tl.store(data0 + 1, val1, mask = True) 
  tl.store(data0 + 2, val2, mask = True) 
  tl.store(data0 + 3, val3, mask = True) 
  ```
  
 to explicit cast:
 ```
@triton.jit                                                                                                                             
def E_4(data0,data1):                                                                                                                   
  val0 = tl.load(data1 + 0, mask = True).to(tl.float32)                                                                                 
  val1 = tl.load(data1 + 1, mask = True).to(tl.float32)                                                                                 
  val2 = tl.load(data1 + 2, mask = True).to(tl.float32)                                                                                 
  val3 = tl.load(data1 + 3, mask = True).to(tl.float32)                                                                                 
  tl.store(data0 + 0, val0.to(tl.int1), mask = True)                                                                                    
  tl.store(data0 + 1, val1.to(tl.int1), mask = True)                                                                                    
  tl.store(data0 + 2, val2.to(tl.int1), mask = True) 
  tl.store(data0 + 3, val3.to(tl.int1), mask = True) 
  ```